QuestionAnswering/QA_SEQ2SEQ/deployment/fastapi_app/services/translate/sparql/graph_pattern.py

```diff
@@ -0,0 +1,231 @@
+
+    def __init__(self, subj: str, tails: Iterable[Tuple[str, str]]):
+        object.__setattr__(self, "subj", subj)
+        object.__setattr__(self, "tails", tuple(tails))
+
+    @classmethod
+    def from_triple(cls, subj: str, predicate: str, obj: str):
+        return cls(subj=subj, tails=[(predicate, obj)])
+
+    def __str__(self):
+        return "{subj} {tails} .".format(
+            subj=self.subj,
+            tails=" ; ".join(
+                [
+                    "{predicate} {obj}".format(predicate=predicate, obj=obj)
+                    for predicate, obj in self.tails
+                ]
+            ),
+        )
+
+    @classmethod
+    def extract(cls, sparql_fragment: str):
+        subj, sparql_fragment = sparql_fragment.split(maxsplit=1)
+
+        tails = []
+        while True:
+            sparql_fragment = sparql_fragment.strip()
+            predicate, sparql_fragment = sparql_fragment.split(maxsplit=1)
+
+            sparql_fragment = sparql_fragment.strip()
+            if sparql_fragment.startswith('"'):
+                ptr = 1
+                while ptr < len(sparql_fragment) and sparql_fragment[ptr] != '"':
+                    ptr += 1
+                obj = sparql_fragment[: ptr + 1]
+                sparql_fragment = sparql_fragment[ptr + 1 :].strip()
+            elif sparql_fragment.startswith('['):
+                stack = ['[']
+                ptr = 1
+                obj = None
+                while ptr < len(sparql_fragment):
+                    if sparql_fragment[ptr] in "[":
+                        stack.append('[')
+                    elif sparql_fragment[ptr] == "]":
+                        assert stack[-1] == '[', "Unexpected close square bracket: " + sparql_fragment
+                        if len(stack) == 1:
+                            obj = sparql_fragment[:ptr + 1]
+                            sparql_fragment = sparql_fragment[ptr + 1 :].strip()
+                            break
+                        else:
+                            stack.pop()
+                    elif sparql_fragment[ptr] == '"':
+                        ptr += 1
+                        while ptr < len(sparql_fragment) and sparql_fragment[ptr] != '"':
+                            ptr += 1
+                    else:
+                        pass
+                    ptr += 1
+                assert obj is not None
+            else:
+                splits = sparql_fragment.split(maxsplit=1)
+                if len(splits) == 2:
+                    obj, sparql_fragment = splits
+                else:
+                    (obj,) = splits
+                    sparql_fragment = ""
+
+                if obj.endswith(".") or obj.endswith(";"):
+                    punctuation = obj[-1]
+                    sparql_fragment = punctuation + sparql_fragment
+                    obj = obj[:-1]
+
+            tails.append((predicate, obj))
+
+            assert sparql_fragment[0] in [";", "."], sparql_fragment
```

Comment:
See comment above. Several occurrences in this file.

Reply:
I've replaced the `assert` statement with `raise ValueError(...)`.